src/main/java/org/elasticsearch/cluster/metadata/MetaDataIndexStateService.java

```diff
@@ -160,7 +164,15 @@ public ClusterState execute(ClusterState currentState) {
                 ClusterBlocks.Builder blocksBuilder = ClusterBlocks.builder()
                         .blocks(currentState.blocks());
                 for (String index : indicesToOpen) {
-                    mdBuilder.put(IndexMetaData.builder(currentState.metaData().index(index)).state(IndexMetaData.State.OPEN));
+                    IndexMetaData indexMetaData = IndexMetaData.builder(currentState.metaData().index(index)).state(IndexMetaData.State.OPEN).build();
+                    // The index might be closed because we couldn't import it due to old incompatible version
+                    // We need to check that this index can be upgraded to the current version
+                    try {
+                        indexMetaData = metaDataIndexUpgradeService.upgradeIndexMetaData(indexMetaData);
+                    } catch (Exception ex) {
+                        throw new IndexException(new Index(index), "unpgradable index", ex);
```

Comment:
@kimchy I will change it to "cannot open the index due to upgrade failure" unless you have better suggestions. The reason why this index cannot be upgraded and possible remedies will be in the inner exception that will be returned as part of the response, so there is probably no reason to repeat them in this exception as well.